libopenage/rng/rng.cpp

```diff
@@ -48,58 +49,93 @@ void RNG::seed(uint64_t val) {
- * This function is a good way of mixing up an input,
- * or mixing together two results with hash64(val1 ^ val2)
- *
- * https://en.wikipedia.org/wiki/MurmurHash
- */
-static uint64_t hash64(uint64_t val) {
-	// The two magic numbers have been determined heuristically.
-	// They avalanche all bits in val to within 0.25% bias.
-
-	val ^= val >> 33;
-	val *= 0xff51afd7ed558ccd;
-	val ^= val >> 33;
-	val *= 0xc4ceb9fe1a85ec53;
-	return val ^ (val >> 33);
+* Rotate left no carry
+*
+* http://blog.regehr.org/archives/1063
+*/
+inline uint64_t rotl64(uint64_t x, uint64_t n)
+{
+	assert (n<64);
+	return (x<<n) | (x>>(-n&63));
 }
 
-
-/*
- * reinterpret_casts 8 chars from a buffer as a uint64_t.
- */
-static inline uint64_t from_cptr(const char *data) {
-	uint64_t result;
-	memcpy(reinterpret_cast<char *>(&result), data, sizeof(result));
-	return result;
+// Maybe this should be a lambda inside of siphash()?
+inline void sipround(uint64_t *v) {
+	v[0] += v[1];
+	v[1] = rotl64(v[1], 13);
+	v[1] ^= v[0];
+	v[0] = rotl64(v[0], 32);
+	v[2] += v[3];
+	v[3] = rotl64(v[3], 16);
+	v[3] ^= v[2];
+	v[0] += v[3];
+	v[3] = rotl64(v[3], 21);
+	v[3] ^= v[0];
+	v[2] += v[1];
+	v[1] = rotl64(v[1], 17);
+	v[1] ^= v[2];
+	v[2] = rotl64(v[2], 32);
 }
 
 
-/*
- * hashes an initial seed into the base state
- */
-static void init_hash(uint64_t *state, uint64_t seed) {
-	// Mix the initial seed into data for each state value
-	state[0] = hash64(seed);
-	state[1] = hash64(seed ^ state[0]);
-}
+// I don't know where to put this, or what value should we use
+static const uint64_t siphash_key[2]{0x0706050403020100ull, 0x0f0e0d0c0b0a0908ull};
 
 
 /*
- * For each 8-byte block after the initial seed,
- * mix it with state[0] and set state[0] equal to the result.
- * Then, set state[1] to the mixture of state[1] and state[0]
- * This lets each bit from the input play a role in the output
- * w/o requiring that the size be a multiple of 8 or having extra
- * code for the non multiple-of-8 sections
+ * Siphash
+ *
+ * https://131002.net/siphash/
  */
-static void hash_bytes(uint64_t *state, const char *data, size_t count) {
-	for (size_t i = 1; i <= count; i++) {
-		state[0] = hash64(state[0] ^ from_cptr(data+i));
-		state[1] = hash64(state[0] ^ state[1]);
+uint64_t siphash(const uint64_t *key, const uint8_t *data, size_t len_data) {
+
+  // (len_data mod 256) to append at the end of the data
+	uint64_t ending = (len_data & 0xff) << 56;
+
+	// Initialization
+
+	uint64_t v[4];
+	v[0] = key[0] ^ 0x736f6d6570736575ull;
+	v[1] = key[1] ^ 0x646f72616e646f6dull;
+	v[2] = key[0] ^ 0x6c7967656e657261ull;
+	v[3] = key[1] ^ 0x7465646279746573ull;
+
+	// Compression
+
+	{
+		size_t rem_bytes{len_data};
+		const uint8_t *idx{data}; // Points to the start of the current block being copied
+		uint64_t block;
+		bool finished{false};
+
+		do {
+			block = 0;
+			if (rem_bytes < 8) {
+				finished = true;
+				memcpy(reinterpret_cast<uint8_t *>(&block), idx, rem_bytes);
```

Comment:
I'm thinking about how it could be possible to work read-only on the input data and just get back the result then (to avoid `memcpy`)